src/reportHelpers.ts

```diff
@@ -0,0 +1,228 @@
+import { Map as IMap, List, fromJS, Collection } from 'immutable';
+import { BigNumber } from 'bignumber.js';
+
+import { TaxReport } from './types';
+import TaxLot from './taxLot';
+import Disposal from './disposal';
+import moment from 'moment';
+
+export const applyUnmatchedSales = (report: TaxReport): IMap<any, any> => {
+  let currentReport = report;
+  report.reduce(
+    (
+      aggregatePriorSales: IMap<any, any>,
+      reportYear: IMap<any, any>,
+      year: string
+    ): IMap<any, any> => {
+      const unmatched = reportYear.get('unmatched');
+
+      if (unmatched.size > 0) {
+        const currentAndPriorSales = unmatched.reduce(
+          (red: IMap<any, any>, unmatchedSale: IMap<any, any>) => {
+            const existing = red.get(unmatchedSale.get('asset'), new BigNumber(0));
+            return red.set(
+              unmatchedSale.get('asset'),
+              BigNumber.sum(existing, unmatchedSale.get('asset_amount'))
+            );
+          },
+          aggregatePriorSales
+        );
+
+        // reduce total holdings per asset in report
+        currentAndPriorSales.map((amount: BigNumber, holdingAsset: string) => {
+          const keypath = [year, 'assets', holdingAsset, 'holdings'];
+          const existingHoldings = currentReport.getIn(keypath, new BigNumber(0));
+          currentReport = currentReport.setIn(keypath, existingHoldings.minus(amount));
+        });
+
+        return currentAndPriorSales;
+      } else {
+        aggregatePriorSales.map((amount: BigNumber, holdingAsset: string) => {
+          const keypath = [year, 'assets', holdingAsset, 'holdings'];
+          const existingHoldings = currentReport.getIn(keypath, new BigNumber(0));
+          currentReport = currentReport.setIn(keypath, existingHoldings.minus(amount));
+        });
+        return aggregatePriorSales;
+      }
+    },
+    IMap({})
+  );
+
+  return currentReport;
+};
+
+export const addIncomeToReport = (report: IMap<any, any>, taxLot: TaxLot): IMap<any, any> => {
```

Comment:
Ditto on updating IMAP types. Return type is pretty clear at end of func

Reply:
done